### ISSUE TITLE
Blurry Font rendering using Terminal Services/Citrix

### ISSUE THREAD

**Issue by Amal22** [cefsharp/CefSharp#1350]
Hello,

We have successfully embedded CefSharp in a .Net application, version 41.0.0. Running the application on terminal server, we have encountred a problem with font rendering : fonts seem to be degraded in terminal server. We haven't encounter this problem on local machines. The picture below describes the  behaviour of Chromium on terminal server: 
![problem-font-rendering-chromium](https://cloud.githubusercontent.com/assets/13152445/10590949/7e337552-76ab-11e5-93df-5c3852b60e5f.png)

Do you have any idea?

**Comment by amaitland** [comment]
Try some of the command line arguments, like disable directwrite, also you might need to disable gpu explicitly.

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Example/CefExample.cs#L74

**Comment by Amal22** [comment]
Ok, thank you for the reply @amaitland :) I'll try it !

**Comment by svantreeck** [comment]
We ran into a similar issue on citrix and found it was related to how fonts get rendered in chrome with font smoothing disabled and direct write enabled...our workaround was to disable direct write to avoid aliasing in the font rendering.

```
    if (!NativeMethods.GetFontSmoothing())
    {
        settings.CefCommandLineArgs["-disable-direct-write"] = "true";
    }

    Cef.Initialize(settings, true, true);

////

public class NativeMethods
    {
        [DllImport("user32.dll", SetLastError = true)]
        static extern bool SystemParametersInfo(uint uiAction, uint uiParam, ref int pvParam, uint fWinIni);

        /* Constants used for User32 calls. */
        const uint SPI_GETFONTSMOOTHING = 74;

        public static Boolean GetFontSmoothing()
        {
            bool iResult;
            int pv = 0;
            /* Call to systemparametersinfo to get the font smoothing value. */
            iResult = SystemParametersInfo(SPI_GETFONTSMOOTHING, 0, ref pv, 0);
            if (pv > 0)
            {
                //pv > 0 means font smoothing is on.
                return true;
            }
            else
            {
                //pv == 0 means font smoothing is off.
                return false;
            }
        }
    }
```

**Comment by amaitland** [comment]
Looks like a reasonable solution. If your up for creating a PR then this could be added to the CefSharp example project.

**Comment by amaitland** [comment]
@Amal22 How'd you go? Happy to close this?

**Comment by Amal22** [comment]
Sorry, I took a time to respond.
I tried the solution of @svantreeck. But, we still have the problem !

**Comment by Amal22** [comment]
Sorry, I took a time to respond.
I tried the solution of @svantreeck. But, we still have the problem !

**Comment by amaitland** [comment]
If you just explicitly use `settings.CefCommandLineArgs.Add("disable-direct-write", "1");` and ignore the rest of it for now, does that work?

**Comment by Amal22** [comment]
Yes, I tried it. It doesn't go !

**Comment by amaitland** [comment]
Did you also disable gpu?

**Comment by Amal22** [comment]
Yes, I tried both : settings.CefCommandLineArgs.Add("disable-direct-write", "1")
 and settings.CefCommandLineArgs.Add("disable-gpu", "1"). But, we still have the same problem.

**Comment by jankurianski** [comment]
- Does the problem occur with other Terminal Services servers, or just the one server?
- What OS is the Terminal Services server running that has this problem?
- Does the same version of CEF have the same font problems on the same server?
- Does the same version of Google Chrome have the same font problems on the same server?
- Have you tried a newer version of CefSharp?
- I don't see where you said whether you were using WPF or WinForms?

i have used CefSharp 43 and 45 in Remote Desktop sessions with Windows 7/10, and also in Terminal Services on Windows Server 2012 with no font issues.

**Comment by amaitland** [comment]
> i have used CefSharp 43 and 45 in Remote Desktop sessions with Windows 7/10, and also in Terminal Services on Windows Server 2012 with no font issues.

Thanks for the feedback @jankurianski :+1:

**Comment by Amal22** [comment]
* I don't see where you said whether you were using WPF or WinForms?
=> Actually, I am using CefSharp Winforms version 41 in a CAB application  (Composite UI Application Block). We integrated the web browser in many windows. When the user log in the application, the view is loaded and Chromium is initialized.
This is where I initialized Chromium:

```
 public void RepositoryConnect()
        {
            connectionThread = new BackgroundWorker();
            connectionThread.RunWorkerCompleted += new RunWorkerCompletedEventHandler(connectionThread_RunWorkerCompleted);
            connectionThread.DoWork += new DoWorkEventHandler(connectionThread_DoWork);
            LoginData data = new LoginData() { UserId = View.GetUserId(), UserPassword = View.GetUserPassword(), ForceConnection = View.ForceUserConnection, SelectedRepository = View.GetSelectedRepository() };

            //Hook to debug the Unhandled exception window
            if (data.UserId == "EXCEPTION")
            {
                throw new Exception("TEST EXCEPTION MESSAGE");

            }

            connectionThread.RunWorkerAsync(data); ;
        }
```
///////////////////////
```

 void connectionThread_DoWork(object sender, DoWorkEventArgs e)

{
    //some code
    try
    {
        ChromiumInitializer.Initialize(UserPreferences.ConnectionLanguage);
    }
    catch (Exception ex)
    {
        MessageBox.Show("An error occured when initialising chromium");
        System.Windows.Forms.Application.Exit();
    }
    //some code
}
```
* What OS is the Terminal Services server running that has this problem?

=> The OS installed is Windows Server 2008 R2.

* Does the problem occur with other Terminal Services servers, or just the one server?

We run the application on windows 7/8 and in Terminal Services On  Windows Server 2008 R2. We didn't run it on another server.

* When we run the .Net application in local machines (on Windows 7/8), fonts are OK.
* When we run the application in Terminal Services on Windows Server 2008 R2, the fonts are blurry.

* Have you tried a newer version of CefSharp?
=> No, I didn't try a newer version of CefSharp.

* We installed Google Chrome on Windows Server 2008 => Fonts are OK

* I implemented another winforms application where I used the same CefSharp version. This application is composed of a simple form.
 ```
 public partial class Form1 : Form
    {
        private ChromiumWebBrowser _webBrowser;
        
        public Form1()
        {
            InitializeComponent();
            ChromiumInitializer.Initialize("fr-FR");
        }
        private void Form1_Load(object sender, EventArgs e)
        {
           _webBrowser = new ChromiumWebBrowser(string.Empty);
            _webBrowser.Dock = DockStyle.Fill;
            panel1.Controls.Add(_webBrowser);

            while (!_webBrowser.IsBrowserInitialized)
            {
                Thread.Sleep(200);
                Application.DoEvents();
            }
            _webBrowser.Load("file:///web/script_preview.html");
        }
   } 
 public static class ChromiumInitializer
    {
        public static void Initialize(string language)
        {
            var settings = new CefSettings();
            switch (language)
            {
                case "en-US":
                    settings.Locale = "en-US";
                    break;
                case "fr-FR":
                    settings.Locale = "fr";
                    break;

            }
            settings.RegisterScheme(new CefCustomScheme()
            {
                SchemeName = LocalSchemeHandlerFactory.SchemeName,
                SchemeHandlerFactory = new LocalSchemeHandlerFactory()
            });

            if (!NativeMethods.GetFontSmoothing())
            {
                settings.CefCommandLineArgs["disable-direct-write"] = "true";
            }
            settings.LogSeverity = LogSeverity.Disable;
            Cef.Initialize(settings, true, true);
        }
    }
```
When I run this application in terminal services, I didn’t encounter font problems.
I think this depends on where we initialize Chromium in the application.

**Comment by amaitland** [comment]
You appear to be initializing on a background worker thread. You need to call `Cef.Initialize` on your main applicaiton thread, e.g. the `UI` thread. If you don't explicitly call `Cef.Initialize` the constructor of the first `ChromiumWebBrowser` will call `Cef.Initialize` for you.

https://htmlpreview.github.io/?https://github.com/cefsharp/CefSharp/blob/cefsharp/45/CefSharp.Example/Resources/Home.html#the-basics

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/Cef.h#L161

**Comment by jankurianski** [comment]
@amaitland is definitely right about that... there could be a timing issue where the `ChromiumWebBrowser` calls `Initialize` first and your custom `Initialize` is ignored.

Apart from that, there is also a high risk your  application will freeze or crash when shutting down.  It's happened to me before when I incorrectly initialized on background threads (or called `Cef.Shutdown` on background threads).

**Comment by amaitland** [comment]
> there could be a timing issue where the ChromiumWebBrowser calls Initialize first and your custom Initialize is ignored.

@jankurianski Excellent point!

**Comment by Amal22** [comment]
Ok, I will call ```Cef.Initialize``` on the main application and we will see if that solves the problem.

**Comment by amaitland** [comment]
> Me too, I have a doubt about that.

Doubt about what exactly?

**Comment by Amal22** [comment]
I have a doubt that the problem is due to initializing Chromium on a background thread.

**Comment by amaitland** [comment]
> I have a doubt that the problem is due to initializing Chromium on a background thread.

The `CEF` documentation is fairly explicit about initializing on the main thread. 

http://magpcss.org/ceforum/apidocs3/projects/%28default%29/%28_globals%29.html#CefInitialize%28constCefMainArgs&,constCefSettings&,CefRefPtr%3CCefApp%3E,void*%29

**Comment by amaitland** [comment]
Also with `41.0.0` multiple calls to `Cef.Initialize` are ignored, so if you for instance creating an instance of `ChromiumWebBrowser`, then call `Cef.Initialize`, the call to `Cef.Initailize` will be ignored as the `ChromiumWebBrowser` instance will have already performed the initialization.

In `43.0.x` versions and above a second call to `Cef.Initialize` will throw an exception. There have been quite a few instances where people have had the order incorrect and it caused issues.

**Comment by amaitland** [comment]
Another handy check is to load `chrome://version/` and see what command line params were used (You can also use task manager and inspect the `BrowserSubProcess.exe` with `--type=render` command line args to see if they're being passed correctly.

**Comment by Amal22** [comment]
I used these settings: 
```
settings.CefCommandLineArgs.Add("disable-gpu", "1");
settings.CefCommandLineArgs.Add("disable-direct-write", "1");
```
When I loaded ```chrome://version/``` in the web browser, I found that the filters are present.
```
Command Line	"...
 --browser-subprocess-path=CefSharp.BrowserSubprocess.exe --no-sandbox --lang=fr --log-severity=disable --disable-gpu=1 --disable-direct-write=1
```

**Comment by amaitland** [comment]
After you've changed how the `Initialize

Can you try the latest `MinimalExample` with direct write disabled.

https://github.com/cefsharp/CefSharp.MinimalExample

**Comment by amaitland** [comment]
Have you changed how `Cef.Initialize` is called out of curiosity? The fact that you said it works in your other `CefSharp` based application is interesting.

If changing how you call initialize doesn't work then can you try the latest `MinimalExample` with direct write disabled.

https://github.com/cefsharp/CefSharp.MinimalExample

**Comment by Amal22** [comment]
Hello, 
I met ``` Cef.Initialize ``` in the main method. But, unfortunately this doesn't solve the problem. The code is here:
```
[STAThread]
        [HandleProcessCorruptedStateExceptions]
        private static void Main(string[] args)
        {
            try
            {
                ChromiumInitializer.Initialize("fr-FR");
            }
            catch (Exception ex)
            {
                MessageBox.Show(Module.ResourceManagerService.GetString("error_initialising_chromium") + " : " + ex.Message);
                System.Windows.Forms.Application.Exit();
            }
            Infragistics.Win.DrawUtility.UseGDIPlusTextRendering = false;
#if (DEBUG)
            RunInDebugMode(args);
#else
                RunInReleaseMode(args);
#endif
        }

        private static void RunInDebugMode(params string[] args)
        {
            Application.SetCompatibleTextRenderingDefault(false);
            SimpleUnhandledExceptionManager.AddHandler(false);

            ShellApplication shellApplication = new ShellApplication();
            if (args != null && args.Length > 0)
                shellApplication._autoLogin = GetAutoLogin(args);
            shellApplication.Run();
        }

        private static void RunInReleaseMode(params string[] args)
        {
            //AppDomain.CurrentDomain.UnhandledException += new UnhandledExceptionEventHandler(AppDomainUnhandledException);
            SimpleUnhandledExceptionManager.AddHandler(false);

            Application.SetCompatibleTextRenderingDefault(false);

            ShellApplication shellApplication = new ShellApplication();
            if (args != null && args.Length > 0)
                shellApplication._autoLogin = GetAutoLogin(args);
            shellApplication.Run();
        }
```
In the main method, GDI+ is disabled. Can this be related to the problem?
 I search in the wiki of Chromium and I found that Chromium uses Skia for nearly all graphics operations, including text rendering. GDI is for the most part only used for native theme rendering.
https://www.chromium.org/developers/design-documents/graphics-and-skia

**Comment by jankurianski** [comment]
> In the main method, GDI+ is disabled. Can this be related to the problem?

If you remove the line where GDI+ is disabled, does the problem go away?

**Comment by Amal22** [comment]
I removed the line ```Infragistics.Win.DrawUtility.UseGDIPlusTextRendering = false; ``` and the problem doesn't go away.

**Comment by jankurianski** [comment]
Unfortunately for you, it seems that the problem is unique to your application.  This makes it impossible for us to help.  I recommend you start with the CefSharp minimal example project, and keep adding code from your project to the minimal example, until it starts to exhibit the same problem.

**Comment by amaitland** [comment]
@Amal22 Did you have any luck tracking down your problem?

**Comment by Amal22** [comment]
Hello,
We solved this problem a few days ago. Sorry, I didn't tell you the solution because I have much work to do.
In our application, to avoid code redundancy, we created a user control named "ChromiumUserControl" where we integrated the Chromium web browser and implemented the various functions that we use. So, to integrate Chromium in a view, instead of creating an instance of ChromiumWebBrowser , we create an instance of "ChromiumUserControl".
To solve the font rendering problem, we disabled automatic scaling of "ChromiumUserControl" by setting autoScaleMode property to "None". We also enabled ClearType on Windows Server 2008.
